titus-server-federation/src/main/java/com/netflix/titus/federation/service/FallbackJobServiceGateway.java

```diff
@@ -1,64 +1,254 @@
+    }
+
+    @Override
+    public Observable<Task> findTask(String taskId, CallMetadata callMetadata) {
+        final String methodName = "findTask";
+        Observable<Task> primaryObservable = primary.findTask(taskId, callMetadata);
+        Observable<Task> secondaryObservable = secondary.findTask(taskId, callMetadata);
+        return getFallbackObservable(methodName, primaryObservable, secondaryObservable);
+    }
+
+    @Override
+    public Observable<TaskQueryResult> findTasks(TaskQuery taskQuery, CallMetadata callMetadata) {
+        final String methodName = "findTasks";
+        Observable<TaskQueryResult> primaryObservable = primary.findTasks(taskQuery, callMetadata);
+        Observable<TaskQueryResult> secondaryObservable = secondary.findTasks(taskQuery, callMetadata);
+        return getFallbackObservable(methodName, primaryObservable, secondaryObservable);
+    }
+
+    @Override
+    public Completable killTask(TaskKillRequest taskKillRequest, CallMetadata callMetadata) {
+        final String methodName = "killTask";
+        Completable primaryCompletable = primary.killTask(taskKillRequest, callMetadata);
+        Completable secondaryObservable = secondary.killTask(taskKillRequest, callMetadata);
+        return getFallbackCompletable(methodName, primaryCompletable, secondaryObservable);
+    }
+
+    @Override
+    public Completable updateTaskAttributes(TaskAttributesUpdate request, CallMetadata callMetadata) {
+        final String methodName = "updateTaskAttributes";
+        Completable primaryCompletable = primary.updateTaskAttributes(request, callMetadata);
+        Completable secondaryObservable = secondary.updateTaskAttributes(request, callMetadata);
+        return getFallbackCompletable(methodName, primaryCompletable, secondaryObservable);
+    }
+
+    @Override
+    public Completable deleteTaskAttributes(TaskAttributesDeleteRequest request, CallMetadata callMetadata) {
+        final String methodName = "deleteTaskAttributes";
+        Completable primaryCompletable = primary.deleteTaskAttributes(request, callMetadata);
+        Completable secondaryObservable = secondary.deleteTaskAttributes(request, callMetadata);
+        return getFallbackCompletable(methodName, primaryCompletable, secondaryObservable);
+    }
+
+    @Override
+    public Completable moveTask(TaskMoveRequest taskMoveRequest, CallMetadata callMetadata) {
+        final String methodName = "moveTask";
+        Completable primaryCompletable = primary.moveTask(taskMoveRequest, callMetadata);
+        Completable secondaryObservable = secondary.moveTask(taskMoveRequest, callMetadata);
+        return getFallbackCompletable(methodName, primaryCompletable, secondaryObservable);
+    }
+
+    private <T> Observable<T> getFallbackObservable(String methodName, Observable<T> primary, Observable<T> secondary) {
+        if (federationConfiguration.isRemoteFederationEnabled()) {
+            return primary.onErrorResumeNext( (t) -> {
```

Comment:
For operations with side effects, we run the risk of retrying un-retryable errors here, i.e. on partial failures where the remote service partially succeeded but the call returns an error - e.g. a timeout.

This is OK for a first iteration, but did you consider limiting what types of errors cause a fallback to `secondary`? Perhaps only `UNIMPLEMENTED` should trigger a fallback?

Reply:
This is a good idea `UNIMPLEMENTED` and `UNAVAILABLE` would probably be good choices.  Let me think about it a little.

Reply:
sgtm, I'd caution against `UNAVAILABLE` as well, I'm not sure it could be thrown for in-flight requests. Perhaps I'm being a bit too paranoid.